{"cells":[{"cell_type":"markdown","metadata":{},"source":["# Pangenomics workflow\n","\n","## Background\n","\n","This notebook contains the steps to create a pangenome for 9 *Escherichia*/*Shigella* type strains listed in Table 1.  \n","This workflow uses [Anvi'o](https://anvio.org/) to calculate and visualise the pangenome.  \n","\n","**Table 1.** Type strains for pangenomics analysis. Shiga-toxin producing *E.coli* in **bold**.  \n","\n","| Assembly Accession | Organism Name | Total Sequence Length | Assembly Level | Assembly Submission Date |\n","|------------------|-------------|--------------------:|--------------|----------------------:|\n","| GCF_000005845.2 | *Escherichia coli* str. K-12 substr. MG1655 | 4641652 | Complete Genome | 2013-09-26 |\n","| GCF_000006925.2 | *Shigella flexneri* 2a str. 301 | 4828820 | Complete Genome | 2011-08-03 |\n","| **GCF_000008865.2** | ***Escherichia coli*** **O157:H7 str. Sakai** | **5594605** | **Complete Genome** | **2018-06-08** |\n","| GCF_002290485.1 | *Shigella boydii* | 4825405 | Contig | 2017-09-12 |\n","| GCF_013374815.1 | *Shigella sonnei* | 4762774 | Complete Genome | 2020-06-27 |\n","| GCF_016904755.1 | *Escherichia albertii* | 4631903 | Complete Genome | 2021-03-10 |\n","| GCF_020097475.1 | *Escherichia fergusonii* | 4645928 | Complete Genome | 2021-09-23 |\n","| GCF_022354085.1 | *Shigella dysenteriae* | 5192674 | Complete Genome  | 2022-02-22 |\n","| GCF_902709585.1 | *Escherichia marmotae* | 4584809 | Contig | 2020-07-20 |"]},{"cell_type":"markdown","metadata":{},"source":["\n","## Downloading data\n","\n","We will use NCBI Datasets command-line tool to download each genome using the assembly accessions from the table above. We only need the annotated genbank files (`gbff`), which we can specify in the command. "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","\n","datasets download genome accession \\\n","        GCF_000005845.2 GCF_000006925.2 GCF_000008865.2 \\\n","        GCF_002290485.1 GCF_013374815.1 GCF_016904755.1 \\\n","        GCF_020097475.1 GCF_022354085.1 GCF_902709585.1 \\\n","        --include gbff \\\n","        --no-progressbar"]},{"cell_type":"markdown","metadata":{},"source":["The assemblies are downloaded in one compressed folder, which needs to be unpacked."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","\n","unzip ncbi_dataset.zip -d escherichia_shigella"]},{"cell_type":"markdown","metadata":{},"source":["## File formatting\n","\n","Each annotated assembly file is stored in separatae folder. We need to keep this in mind when giving the paths to each assembly file.  \n","Next step is to format the genbank files into files that anvi'o needs for calculating the pangenome.  \n","We'll use a script called `anvi-script-process-genbank` that splits the genbank file into three different files:\n"," - `*-contigs.fasta`: file with all the contigs\n"," - `*-gene-calls.txt`: file with coordinates for each gene call in the contigs\n"," - `*-functions.txt`: functional annotation for each gene call \n","\n","As said, the genbank files are in separate folders and they all have the same name (`genomic.gbff`). But the folder name is the accession number. So we need to use the folder name as our genome name. The beginning of the code below first removes the common path to all genbank folders (`escherichia_shigella/ncbi_dataset/data/`) and then removes the redundant file name (`genomic.gbff`). This will leave us with the assembly accession. "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","\n","for genome in `ls escherichia_shigella/ncbi_dataset/data/*/genomic.gbff`\n","do\n","    name=${genome#escherichia_shigella/ncbi_dataset/data/} \n","    name=${name%/genomic.gbff} \n","    \n","    anvi-script-process-genbank \\\n","        -i $genome \\\n","        --output-fasta escherichia_shigella/ncbi_dataset/data/${name}/${name}-contigs.fasta \\\n","        --output-gene-calls escherichia_shigella/ncbi_dataset/data/${name}/${name}-gene-calls.txt \\\n","        --output-functions escherichia_shigella/ncbi_dataset/data/${name}/${name}-functions.txt \\\n","        --annotation-source prodigal \\\n","        --annotation-version v2.6.3\n","done"]},{"cell_type":"markdown","metadata":{},"source":["## Build the pangenome\n","\n","After splitting the genbank files we need a file (`fasta.txt`) for anvi'o that has the names of the genomes and the paths for all the files for each of our genomes.  \n","The code below goes thru all the files produced by `anvi-script-process-genbank` and prints their paths to the `fasta.txt` file. The first line just prints the column headers. "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash \n","\n","echo -e \"name\\tpath\\texternal_gene_calls\\tgene_functional_annotation\" > fasta.txt\n","for strain in `ls escherichia_shigella/ncbi_dataset/data/*/*-contigs.fasta`\n","do\n","    strain_name=${strain#escherichia_shigella/ncbi_dataset/data/*/}\n","    echo -e ${strain_name%-contigs.fasta}\"\\t\"$strain\"\\t\"${strain%-contigs.fasta}\"-gene-calls.txt\\t\"${strain%-contigs.fasta}\"-functions.txt\"\n","done >> fasta.txt"]},{"cell_type":"markdown","metadata":{},"source":["\n","We will also need a configuration file called `config.json` that has all the options we want to use while building the pangenome.  \n","You do not have to make this file, it is provided.  \n","\n","The content of the configuration file is: \n","```\n","{\n","    \"workflow_name\": \"pangenomics\",\n","    \"config_version\": \"2\",\n","    \"max_threads\": \"8\",\n","    \"project_name\": \"Escherichia_Shigella_pangenome\",\n","    \"external_genomes\": \"external-genomes.txt\",\n","    \"fasta_txt\": \"fasta.txt\",\n","    \"anvi_gen_contigs_database\": {\n","        \"--project-name\": \"{group}\",\n","        \"--description\": \"\",\n","        \"--skip-gene-calling\": \"\",\n","        \"--ignore-internal-stop-codons\": true,\n","        \"--skip-mindful-splitting\": \"\",\n","        \"--contigs-fasta\": \"\",\n","        \"--split-length\": \"\",\n","        \"--kmer-size\": \"\",\n","        \"--skip-predict-frame\": \"\",\n","        \"--prodigal-translation-table\": \"\",\n","        \"threads\": \"\"\n","    },\n","    \"anvi_pan_genome\": {\n","      \"threads\": \"8\"\n","    }\n","}\n","```\n","\n","Then we have all the files needed to construct the pangenome for our 9 strains. We will use the pangenomics workflow from anvi'o.  \n","You can read more about anvi'o workflows and the different options from [anvi'o workflow help pages](https://anvio.org/help/main/programs/anvi-run-workflow/).  "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[" %%bash\n"," \n"," anvi-run-workflow -w pangenomics -c config.json"]},{"cell_type":"markdown","metadata":{},"source":["After this is done, we have only few steps and then we can visualise our pangenome with anvi'o.  \n","First we import a configuration file that anvi'o calls state, just to have some pre-formatting in our visualisation.  "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","\n","anvi-import-state -n default -p 03_PAN/Escherichia_Shigella_pangenome-PAN.db -s default_state.json "]},{"cell_type":"markdown","metadata":{},"source":["Then to make things easier, go to the `03_PAN` folder and print the full path to the folder.  \n","We will use this path to navigate to the right folder in Visual Studio Code in the last step. So copy it from the notebook. "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","\n","cd 03_PAN\n","pwd"]},{"cell_type":"markdown","metadata":{},"source":["## Visualise the pangenome\n","\n","Then we are all done and can open **Visual Studio code**.   \n","Connect to the server and open the folder containing our pangenome files (the path printed above).  \n","Then run the following command, but remember to change the port to your own port. Each student need to have their own port to be able to visualise the pangenome on their own local browser. "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","\n","cd /path/to/course_folder/pangenome_folder\n","\n","module load anvio\n","YOUR_PORT=YOUR_PORT_NUMBER\n","\n","anvi-display-pan -P $YOUR_PORT"]}],"metadata":{"kernelspec":{"display_name":"Python 3.8.12 ('base')","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.8.12"},"orig_nbformat":4,"vscode":{"interpreter":{"hash":"0c1aed64d5e4f622e107948ca29209dcff88f0cd53bde2862acf33e5a35c0f44"}}},"nbformat":4,"nbformat_minor":2}
