{"cells":[{"cell_type":"markdown","metadata":{},"source":["# Pangenomics workflow\n","\n","## Background\n","\n","This notebook contains the steps tto create a pangenome for 11 Escherichia-Shigella type strains listed in Table 1.  \n","This workflow uses [Anvi'o](https://anvio.org/) to calculate and visualise the pangenome.  \n","\n","**Table 1.** Type strains for pangenomics analysis  \n","\n","| Assembly Accession | Organism Name | Total Sequence Length | Assembly Level | Assembly Submission Date |\n","| ------------------ | ------------- | --------------------- | -------------- |  ----------------------- |\n","| GCF_000005845.2 | Escherichia coli str. K-12 substr. MG1655 | 4641652 | Complete Genome | 2013-09-26 |\n","| GCF_000006925.2 | Shigella flexneri 2a str. 301 | 4828820 | Complete Genome | 2011-08-03 |\n","| GCF_000008865.2 | Escherichia coli O157:H7 str. Sakai\t5594605 | Complete Genome | 2018-06-08 |\n","| GCF_002290485.1 | Shigella boydii | 4825405 | Contig | 2017-09-12 |\n","| GCF_013374815.1 | Shigella sonnei | 4762774 | Complete Genome | 2020-06-27 |\n","| GCF_016904755.1 | Escherichia albertii | 4631903 | Complete Genome | 2021-03-10 |\n","| GCF_020097475.1 | Escherichia fergusonii | 4645928 | Complete Genome | 2021-09-23 |\n","| GCF_020283705.1 | Escherichia whittamii | 4563680 | Scaffold | 2021-10-05 |\n","| GCF_022354085.1 | Shigella dysenteriae | 5192674 | Complete Genome  | 2022-02-22 |\n","| GCF_024733345.1 | Escherichia ruysiae| 4893780 | Contig | 2022-08-22 |\n","| GCF_902709585.1 | Escherichia marmotae | 4584809 | Contig | 2020-07-20 |"]},{"cell_type":"markdown","metadata":{},"source":["\n","## Downloading data\n","\n","We will use NCBI Dataset command-line tool to download each genome using the assembly accessions from table above. We only need the annotated genbank files (`gbff`), so we can exclude all other formats available for each assembly. "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","datasets download genome accession \\\n","        GCF_000005845.2,GCF_000006925.2,GCF_000008865.2,\\\n","        GCF_002290485.1,GCF_013374815.1,GCF_016904755.1,\\\n","        GCF_020097475.1,GCF_020283705.1,GCF_022354085.1, \\\n","        GCF_024733345.1,GCF_902709585.1 \\\n","        --include-gbff \\\n","        --exclude-protein \\\n","        --exclude-rna \\\n","        --exclude-genomic-cds \\\n","        --exclude-seq \\\n","        --exclude-gff"]},{"cell_type":"markdown","metadata":{},"source":["The assemblies are downloaded in one folder in  compressed format, which needs to be unpacked."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","\n","unzip ncbi_dataset.zip -d escherichia_shigella"]},{"cell_type":"markdown","metadata":{},"source":["## File formatting\n","\n","Each annotated assembly file is stored in separatae folder. We need to keep this in mind when giving the paths to each aassembly file.  \n","Next step is to format the genbank files into three separate files that can be used by the "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","for genome in $(ls escherichia_shigella/ncbi_dataset/data/*/genomic.gbff)\n","do\n","    name=${genome#escherichia_shigella/ncbi_dataset/data/} \n","    name=${name%/genomic.gbff} \n","    anvi-script-process-genbank \\\n","        -i $genome \\\n","        --output-fasta escherichia_shigella/ncbi_dataset/data/${name}/${name}-contigs.fasta \\\n","        --output-gene-calls escherichia_shigella/ncbi_dataset/data/${name}/${name}-gene-calls.txt \\\n","        --output-functions escherichia_shigella/ncbi_dataset/data/${name}/${name}-functions.txt \\\n","        --annotation-source prodigal \\\n","        --annotation-version v2.6.3\n","done"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash \n","\n","echo -e \"name\\tpath\\texternal_gene_calls\\tgene_functional_annotation\" > fasta.txt\n","for strain in $(ls escherichia_shigella/ncbi_dataset/data/*/*-contigs.fasta)\n","do\n","    strain_name=${strain#ncbi_dataset/data/*/}\n","    echo -e ${strain_name%-contigs.fasta}\"\\t\"$strain\"\\t\"${strain%-contigs.fasta}\"-gene-calls.txt\\t\"${strain%-contigs.fasta}\"-func\n","tions.txt\"\n","done >> fasta.txt   "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[" %%bash\n"," \n"," anvi-run-workflow -w pangenomics -c config.json"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["%%bash\n","\n","ANVIPORT=YOUR_PORT_NUMBER\n","\n","anvi-display-pan \\\n","    -g 03_PAN/*-GENOMES.db \\\n","    -p 03_PAN/*-PAN.db \\\n","    --server-only -P $ANVIPORT"]}],"metadata":{"kernelspec":{"display_name":"Python 3.8.12 ('base')","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.8.12"},"orig_nbformat":4,"vscode":{"interpreter":{"hash":"0c1aed64d5e4f622e107948ca29209dcff88f0cd53bde2862acf33e5a35c0f44"}}},"nbformat":4,"nbformat_minor":2}
